### PR TITLE
unistd/getcwd: enhance getcwd when buf is NULL

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
Enhance getcwd.
Reference: https://pubs.opengroup.org/onlinepubs/007904975/functions/getcwd.html
## Testing
daily test

### CODE REVIEW

File: libs/libc/unistd/lib_getcwd.c [hunk 1]
```diff
@@ -63,6 +63,12 @@
  *   symbolic links. The 'size' argument is the size in bytes of the
  *   character array pointed to by the 'buf' argument.
  *
+ *   As an extension to the POSIX.1-2001 standard, getcwd() allocates
+ *   the buffer dynamically using malloc if buf is NULL.  In this case,
+ *   the allocated buffer has the length size  unless size is zero, when buf
```
Comment (davids5) [inline libs/libc/unistd/lib_getcwd.c:68]: missing comma?

File: libs/libc/unistd/lib_getcwd.c [hunk 2]
```diff
@@ -92,13 +98,18 @@ FAR char *getcwd(FAR char *buf, size_t size)
...
 #ifdef CONFIG_DEBUG_FEATURES
-  if (!buf || !size)
+  if (buf && !size)
     {
       set_errno(EINVAL);
       return NULL;
     }
 #endif
 
+  if (!size)
```
Comment (davids5) [inline libs/libc/unistd/lib_getcwd.c:108]:
```suggestion
  if (size == 0)
```

File: libs/libc/unistd/lib_getcwd.c [hunk 3]
```diff
@@ -92,13 +98,18 @@ FAR char *getcwd(FAR char *buf, size_t size)
   /* Verify input parameters */
 
 #ifdef CONFIG_DEBUG_FEATURES
-  if (!buf || !size)
+  if (buf && !size)
```
Comment (davids5) [inline libs/libc/unistd/lib_getcwd.c:101]: hmm why does is original code qualified by CONFIG_DEBUG_FEATURES?

File: libs/libc/unistd/lib_getcwd.c (hunk 2 above)
Comment (Donny9) [reply libs/libc/unistd/lib_getcwd.c:108]: Done!

File: libs/libc/unistd/lib_getcwd.c (hunk 1 above)
Comment (Donny9) [reply libs/libc/unistd/lib_getcwd.c:68]: what?

File: libs/libc/unistd/lib_getcwd.c (hunk 3 above)
Comment (Donny9) [reply libs/libc/unistd/lib_getcwd.c:101]: These parameters may be checked under DEBUG? I keep origin layout.

File: libs/libc/unistd/lib_getcwd.c (hunk 1 above)
Comment (davids5) [reply libs/libc/unistd/lib_getcwd.c]: the allocated buffer has the length size  unless size is zero, when buf
the allocated buffer has the length size, unless size is zero, when buf

Comment (davids5) [reply libs/libc/unistd/lib_getcwd.c]: Or remove the extra space?

File: libs/libc/unistd/lib_getcwd.c (hunk 3 above)
Comment (davids5) [reply libs/libc/unistd/lib_getcwd.c]: If it was a error message that makes sense not a return value. The function docs state the return values and is not conditional?  Why not fix it?

File: libs/libc/unistd/lib_getcwd.c (hunk 1 above)
Comment (Donny9) [reply libs/libc/unistd/lib_getcwd.c]: done！

File: libs/libc/unistd/lib_getcwd.c (hunk 3 above)
Comment (Donny9) [reply libs/libc/unistd/lib_getcwd.c]: okay， done！

**Review by davids5 — APPROVED** [review]
@Donny9 - Thank you, the code looks fine now. 

@Donny9  What is the motivation for this change? Is it not to have a `malloc`, `call` `free` pattern? 
I ask because this change seems add marginal value other then encapsulation of the PATH_MAX.  

@patacongo What is you take on this extension?

File: libs/libc/unistd/lib_getcwd.c [hunk 4]
```diff
@@ -115,6 +124,16 @@ FAR char *getcwd(FAR char *buf, size_t size)
       return NULL;
     }
 
+  if (buf == NULL)
+    {
+      buf = malloc(size);
```
Comment (patacongo) [inline libs/libc/unistd/lib_getcwd.c:129]: malloc() should not be used.  Use lib_malloc() instead.  That eliminates the problem case where this function is called from within the OS in certain build modes and malloc does not do what you would want it do do.

**Comment by patacongo** [comment]
> @patacongo What is you take on this extension?

I don't have any strong opinion, as long as it is fully compatible with the require at the OpenGroup.org reference.  We do not want any non-standard behavior in the system.

**Comment by davids5** [comment]
> > @patacongo What is you take on this extension?
> 
> I don't have any strong opinion, as long as it is fully compatible with the require at the OpenGroup.org reference. We do not want any non-standard behavior in the system.

OG says "If buf is a null pointer, the behavior of getcwd() is unspecified."  Are you still good with this once the malloc is fixed?

File: libs/libc/unistd/lib_getcwd.c (hunk 4 above)
Comment (Donny9) [reply libs/libc/unistd/lib_getcwd.c:129]: Okay, thanks. Done!

**Comment by xiaoxiang781216** [comment]
> > > @patacongo What is you take on this extension?
> > 
> > 
> > I don't have any strong opinion, as long as it is fully compatible with the require at the OpenGroup.org reference. We do not want any non-standard behavior in the system.
> 
> OG says "If buf is a null pointer, the behavior of getcwd() is unspecified." Are you still good with this once the malloc is fixed?

Yes, it's glibc specific behaviour. But, OG explicitly state that the behavior is unspecified which mean we can do any action here.
So we still confirm OG specification with this patch, but more compatible with the most used libc implementation.

**Review by davids5 — APPROVED** [review]

**Pull request merged**